src/app-layer-htp.c

```diff
@@ -524,7 +524,7 @@ static uint32_t AppLayerHtpComputeChunkLength(uint64_t content_len_so_far, uint3
 /* below error messages updated up to libhtp 0.5.7 (git 379632278b38b9a792183694a4febb9e0dbd1e7a) */
 struct {
     const char *msg;
-    int  de;
+    uint8_t de;
```

Comment:
should this not be `unsigned int` as the values are defined in an `enum`?

Reply:
An enum value such as `HTTP_DECODER_EVENT_UNKNOWN_ERROR` should be defined as an `int`(signed)
But we can safely cast it to `uint8_t` as we do not have more than 256 values. I even think that this is some kind of preprocessing that acts as `#define` 
Furthermore, we later call `HTPSetEvent` which uses `uint8_t` (because it calls `AppLayerDecoderEventsSetEventRaw` which uses `uint8_t`)

And it does not look like there is warning flag for Wimplicit-enum-int-conversion
cf https://releases.llvm.org/11.0.0/tools/clang/docs/DiagnosticsReference.html#wenum-conversion

So, I think this patch is right

Reply:
hmm I always thought an enum would be an unsigned int, and be 32 bits. But now I'm reading that it may actually be up to the compiler to pick a type as long as it fits? Wonder what that means for using enums in Rust - C - FFI.

Reply:
> Wonder what that means for using enums in Rust - C - FFI.

I guess it means the enums should be defined in Rust, and exported to be used in C with cbindgen :-p

Reply:
> > Wonder what that means for using enums in Rust - C - FFI.
> 
> I guess it means the enums should be defined in Rust, and exported to be used in C with cbindgen :-p

cbindgen outputs the following, I guess in a way to fix the type:
```
enum HTTP2TransactionState {
    HTTP2StateIdle = 0,
    HTTP2StateOpen = 1,
    HTTP2StateReserved = 2,
    HTTP2StateDataClient = 3,
    HTTP2StateHalfClosedClient = 4,
    HTTP2StateDataServer = 5,
    HTTP2StateHalfClosedServer = 6,
    HTTP2StateClosed = 7,
    HTTP2StateGlobal = 8,
};
typedef uint8_t HTTP2TransactionState;
```

So that is one option.

But it looks like the representation of enums has been an issue where the compilers use a different algorithm for layout, like rustc and msvc:

https://github.com/rust-lang/rust/issues/81996

Luckily it looks like rustc, clang, gcc all agree here.

Reply:
So, let's rustify all the things

Reply:
> So, let's rustify all the things

I think it puts us on a path of one big super header file for C.  Not sure if thats really that bad of a thing tho.

Reply:
Could not cbindgen generate one header file per rust module/directory ?

Reply:
> Could not cbindgen generate one header file per rust module/directory ?

No, thats not really how its designed.  You could call it for each file manually and put those into headers.  Then you'd have to worry about ordering of dependent data structures and so on.